### ISSUE TITLE
Scaffold-DbContext throws 'System.InvalidOperationException://SqlServer:Clustered' when reverse engineering a table with clustered hypothetical indexes

### ISSUE THREAD

**Issue by SmithPlatts** [dotnet/efcore#20705]
On a table with a clustered index, and associated hypothetical indexes (statistics), the command `Scaffold-DbContext` throws `System.InvalidOperationException` with the message: _The annotation 'SqlServer:Clustered' cannot be added because an annotation with the same name already exists._

Looking at the code [here](https://github.com/dotnet/efcore/blob/faab5634a5e8b34336f6095f0869a0a2de73c154/src/EFCore.SqlServer/Scaffolding/Internal/SqlServerDatabaseModelFactory.cs#L785), the SQL to reverse engineer Indexes is including hypothetical indexes (`is_hypothetical`).

As per the MS [docuemntation](https://docs.microsoft.com/en-us/sql/relational-databases/system-catalog-views/sys-indexes-transact-sql):

> 1 = Index is hypothetical and cannot be used directly as a data access path. Hypothetical indexes hold column-level statistics.

The workaround for me is to either manually craft the model, or drop the indexes and statistics.

Should the `SqlServerDatabaseModelFactory` be ignoring these statistics though? They cannot be used directly in T-SQL, and clutter up the generated fluent API context (or annotated model).

### Steps to reproduce

Create a table with a hypothetical indexes:

``` sql
CREATE TABLE [dbo].[AdamBugCheck] (
    [MainID] [int] IDENTITY(1,1) NOT FOR REPLICATION NOT NULL
    , [Active] [tinyint] NOT NULL
    , CONSTRAINT [PK_AdamBugCheck] PRIMARY KEY NONCLUSTERED (
        [MainID] ASC
    ) ON [PRIMARY]
) ON [PRIMARY]
GO

CREATE CLUSTERED INDEX [IX_AdamBugCheck_Active] ON [dbo].[AdamBugCheck] (
    [Active] ASC
) WITH (STATISTICS_ONLY = 0) ON [PRIMARY]
GO

CREATE CLUSTERED INDEX [IX_AdamBugCheck_Active_WithStats] ON [dbo].[AdamBugCheck] (
    [Active] ASC
) WITH (STATISTICS_ONLY = -1) ON [PRIMARY]
GO
```

And now try to scaffold it:

``` powershell
Scaffold-DbContext -Connection "Server=mydb.org.tld;Database=MyDB;Integrated Security=True" -Provider Microsoft.EntityFrameworkCore.SqlServer -OutputDir Models -ContextDir . -Tables @("AdamBugCheck") -Verbose
```

You should then be met with:

``` plain
Build started...
Build succeeded.
C:\Program Files\dotnet\dotnet.exe exec --depsfile C:\Repositories\Git\MyDBProj\MyDBProj\bin\Debug\netcoreapp3.1\MyDBProj.deps.json --additionalprobingpath C:\Users\User\.nuget\packages --additionalprobingpath "C:\Program Files\dotnet\sdk\NuGetFallbackFolder" --runtimeconfig C:\Repositories\Git\MyDBProj\MyDBProj\bin\Debug\netcoreapp3.1\MyDBProj.runtimeconfig.json C:\Users\User\.nuget\packages\microsoft.entityframeworkcore.tools\3.1.3\tools\netcoreapp2.0\any\ef.dll dbcontext scaffold "Server=mydb.org.tld;Database=MyDB;Integrated Security=True" Microsoft.EntityFrameworkCore.SqlServer --json --output-dir Models --context-dir . --table AdamBugCheck --force --verbose --no-color --prefix-output --assembly C:\Repositories\Git\MyDBProj\MyDBProj\bin\Debug\netcoreapp3.1\MyDBProj.dll --startup-assembly C:\Repositories\Git\MyDBProj\MyDBProj\bin\Debug\netcoreapp3.1\MyDBProj.dll --project-dir C:\Repositories\Git\MyDBProj\MyDBProj\ --language C# --working-dir C:\Repositories\Git\MyDBProj\MyDBProj --root-namespace MyDBProj
Using assembly 'MyDBProj'.
Using startup assembly 'MyDBProj'.
Using application base 'C:\Repositories\Git\MyDBProj\MyDBProj\bin\Debug\netcoreapp3.1'.
Using working directory 'C:\Repositories\Git\MyDBProj\MyDBProj'.
Using root namespace 'MyDBProj'.
Using project directory 'C:\Repositories\Git\MyDBProj\MyDBProj\'.
Finding design-time services for provider 'Microsoft.EntityFrameworkCore.SqlServer'...
Using design-time services from provider 'Microsoft.EntityFrameworkCore.SqlServer'.
Finding design-time services referenced by assembly 'MyDBProj'.
No referenced design-time services were found.
Finding IDesignTimeServices implementations in assembly 'MyDBProj'...
No design-time services were found.
Found default schema dbo.
Found type alias with name: sys.sysname which maps to underlying data type nvarchar(128).
Found table with name: dbo.AdamBugCheck.
Found column with table: dbo.AdamBugCheck, column name: MainID, ordinal: 1, data type: sys.int, maximum length: 4, precision: 10, scale: 0, nullable: False, identity: True, default value: (null), computed value: (null)
Found column with table: dbo.AdamBugCheck, column name: Active, ordinal: 2, data type: sys.tinyint, maximum length: 1, precision: 3, scale: 0, nullable: False, identity: False, default value: (null), computed value: (null)
Found primary key with name: PK_AdamBugCheck, table: dbo.AdamBugCheck.
Found index with name: IX_AdamBugCheck_Active, table: dbo.AdamBugCheck, is unique: False.
Found index with name: IX_AdamBugCheck_Active_WithStats, table: dbo.AdamBugCheck, is unique: False.
System.InvalidOperationException: The annotation 'SqlServer:Clustered' cannot be added because an annotation with the same name already exists.
   at Microsoft.EntityFrameworkCore.Infrastructure.Annotatable.AddAnnotation(String name, Annotation annotation)
   at Microsoft.EntityFrameworkCore.Infrastructure.Annotatable.AddAnnotation(String name, Object value)
   at Microsoft.EntityFrameworkCore.Infrastructure.Annotatable.Microsoft.EntityFrameworkCore.Metadata.IMutableAnnotatable.AddAnnotation(String name, Object value)
   at Microsoft.EntityFrameworkCore.MutableAnnotatableExtensions.AddAnnotations(IMutableAnnotatable annotatable, IEnumerable`1 annotations)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.RelationalScaffoldingModelFactory.VisitIndex(EntityTypeBuilder builder, DatabaseIndex index)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.RelationalScaffoldingModelFactory.VisitIndexes(EntityTypeBuilder builder, ICollection`1 indexes)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.RelationalScaffoldingModelFactory.VisitTable(ModelBuilder modelBuilder, DatabaseTable table)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.RelationalScaffoldingModelFactory.VisitTables(ModelBuilder modelBuilder, ICollection`1 tables)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.RelationalScaffoldingModelFactory.VisitDatabaseModel(ModelBuilder modelBuilder, DatabaseModel databaseModel)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.RelationalScaffoldingModelFactory.Create(DatabaseModel databaseModel, Boolean useDatabaseNames)
   at Microsoft.EntityFrameworkCore.Scaffolding.Internal.ReverseEngineerScaffolder.ScaffoldModel(String connectionString, DatabaseModelFactoryOptions databaseOptions, ModelReverseEngineerOptions modelOptions, ModelCodeGenerationOptions codeOptions)
   at Microsoft.EntityFrameworkCore.Design.Internal.DatabaseOperations.ScaffoldContext(String provider, String connectionString, String outputDir, String outputContextDir, String dbContextClassName, IEnumerable`1 schemas, IEnumerable`1 tables, Boolean useDataAnnotations, Boolean overwriteFiles, Boolean useDatabaseNames)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.ScaffoldContextImpl(String provider, String connectionString, String outputDir, String outputDbContextDir, String dbContextClassName, IEnumerable`1 schemaFilters, IEnumerable`1 tableFilters, Boolean useDataAnnotations, Boolean overwriteFiles, Boolean useDatabaseNames)
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.ScaffoldContext.<>c__DisplayClass0_0.<.ctor>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.<>c__DisplayClass3_0`1.<Execute>b__0()
   at Microsoft.EntityFrameworkCore.Design.OperationExecutor.OperationBase.Execute(Action action)
The annotation 'SqlServer:Clustered' cannot be added because an annotation with the same name already exists.
```

### Further technical details

EF Core version: 3.1.3
Database provider: Microsoft.EntityFrameworkCore.SqlServer
Target framework: .NET Core 3.1
Operating system: Windows 10 1909
IDE: Visual Studio Professional 2019 16.5.4

**Comment by ajcvickers** [comment]
See also #7665

**Comment by ErikEJ** [comment]
@ajcvickers So I fixed this, and the fix disappeared?

**Comment by ErikEJ** [comment]
Happy to fix it again 😳

**Comment by ajcvickers** [comment]
@ErikEJ Not sure if it's exactly the same thing or not--haven't dug that deep.

**Comment by ajcvickers** [comment]
@ErikEJ But always happy for you to fix it anyway!

**Comment by ErikEJ** [comment]
Sadly looks like the exact same bug! Poor/ difficult test coverage, I suppose.

**Comment by ErikEJ** [comment]
Looks like my previous fix was killed after 8 months: https://github.com/dotnet/efcore/commit/746ff59fae3d587269046604ecf964192958c7ec#diff-4919ea781036f7825206b48d367d5c44

**Comment by ErikEJ** [comment]
Looks like my previous fix was killed after 8 months: https://github.com/dotnet/efcore/commit/746ff59fae3d587269046604ecf964192958c7ec#diff-4919ea781036f7825206b48d367d5c44

**Comment by SmithPlatts** [comment]
Thanks for jumping on this so swiftly @ErikEJ and @ajcvickers! 

After submitting the issue, I did come across [#7665](https://github.com/dotnet/efcore/issues/7665) (I did do quite a search before submitting the issue; I swear!) and was sure that it was related.

Based on the info in that issue, all the hypothetical indexes I have found all start with `_dta_*`. Although I am proficient in SQL Server, I am not a member of the DBA team and therefore can't advise against using DTA ...
